src/pages/RFOX/components/Stake/Bridge/hooks/useRfoxBridgeApproval.tsx

```diff
@@ -0,0 +1,211 @@
+      sellAssetAccountNumber,
+    ],
+  )
+
+  const getFeesWithWalletInput = useMemo(
+    () => ({ ...approvalFeesQueryInput, adapter, wallet }),
+    [adapter, approvalFeesQueryInput, wallet],
+  )
+  const approvalFeesQuery = useEvmFees({
+    ...approvalFeesQueryInput,
+    enabled: isGetApprovalFeesEnabled(getFeesWithWalletInput),
+    staleTime: 30_000,
+    // Ensures fees are refetched at an interval, including when the app is in the background
+    refetchIntervalInBackground: true,
+    // Yeah this is arbitrary but come on, Arb is cheap
+    refetchInterval: isGetApprovalFeesEnabled(getFeesWithWalletInput) ? 15_000 : false,
+  })
+
+  const approvalTx = useAppSelector(gs => selectTxById(gs, serializedApprovalTxIndex ?? ''))
+
+  const isApprovalTxPending = useMemo(
+    () =>
+      approvalMutation.isPending ||
+      (approvalMutation.isSuccess && approvalTx?.status !== TxStatus.Confirmed),
+    [approvalMutation.isPending, approvalMutation.isSuccess, approvalTx?.status],
+  )
+
+  const isApprovalTxSuccess = useMemo(
+    () => approvalTx?.status === TxStatus.Confirmed,
+    [approvalTx?.status],
+  )
+
+  // The approval Tx may be confirmed, but that's not enough to know we're ready to bridge
+  // Allowance then needs to be succesfully refetched - failure to wait for it will result in jumpy states between
+  // the time the Tx is confirmed, and the time the allowance is succesfully refetched
+  // This allows us to detect such transition state
+  const isTransitioning = useMemo(() => {
+    // If we don't have a success Tx, we know we're not transitioning
+    if (!isApprovalTxSuccess) return false
+    // We have a success approval Tx, but approval is still required, meaning we haven't re-rendered with the updated allowance just yet
+    if (isApprovalRequired) return true
+
+    // Allowance has been updated, we've finished transitioning
+    return false
+  }, [isApprovalRequired, isApprovalTxSuccess])
+
+  const handleApprove = useCallback(() => {
+    if (!allowanceContract) return Promise.reject('allowanceContract is required')
+
+    return approvalMutation.mutateAsync()
+  }, [allowanceContract, approvalMutation])
+
+  useEffect(() => {
+    if (!allowanceContract) return
+    if (!approvalTx) return
+    if (isApprovalTxPending) return
+
+    queryClient.invalidateQueries(
+      reactQueries.common.allowanceCryptoBaseUnit(
+        confirmedQuote.sellAssetId,
+        allowanceContract,
+        fromAccountId(confirmedQuote.sellAssetAccountId).account,
+      ),
+    )
+  }, [
+    approvalTx,
+    isApprovalTxPending,
+    queryClient,
+    allowanceContract,
+    confirmedQuote.sellAssetId,
+    confirmedQuote.sellAssetAccountId,
+  ])
+
+  return {
+    isApprovalRequired,
+    allowanceQuery,
+    isGetApprovalFeesEnabled: isGetApprovalFeesEnabled(getFeesWithWalletInput),
+    approvalFeesQuery,
+    isApprovalTxPending,
+    isApprovalTxSuccess,
+    handleApprove,
+    isTransitioning,
+  }
```

Comment:
also this guy can use some memoization

Reply:
https://github.com/shapeshift/web/pull/7079#discussion_r1634369672